Lombiq.NodeJs.Extensions/scripts/copy-assets.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * @summary A script to copy files around; expects configuration in assets-to-copy.json or package.json (assetsToCopy).
+ * @description This script is a wrapper around the npm package "copyfiles". It reads its configuration from a file in
+ *              the consuming project named "assets-to-copy.json" or the "assetsToCopy" node in package.json.
+ */
+const { access } = require('fs').promises;
+const path = require('path');
+const util = require('util');
+/* eslint-disable-next-line import/no-unresolved -- ESLint does not know where to find external modules. */
+const copyfiles = util.promisify(require('copyfiles'));
+const getAssetsConfig = require('./get-assets-config');
+
+const verbose = true;
+const defaultFilePattern = '**/*';
+
+// Change to consuming project's directory.
+process.chdir('../..');
+
+function logLine(message) {
+    if (verbose) process.stdout.write(message + '\n');
+}
+
+async function copyFilesFromConfig(assetsConfig) {
+    return Promise.all(assetsConfig
+        .map((assetsGroup) => assetsGroup.sources.map((assetSource) => {
+            // Normalize the relative path to the directory to remove trailing slashes and straighten out any anomalies.
+            const directoryToCopy = path.relative('.', assetSource);
+            const pattern = assetsGroup.pattern || defaultFilePattern;
+
+            logLine(`Copy assets from "${directoryToCopy}" using pattern "${pattern}"...`);
+
+            return access(directoryToCopy)
+                .then(
+                    () => {
+                        const pathPattern = path.join(directoryToCopy, pattern);
+                        const sourceAndTargetPaths = [pathPattern, assetsGroup.target];
+
+                        // We want to copy all files matched by the given pattern into the target folder mirroring the
+                        // source folder structure. This is done by removing the source folder path from the beginning
+                        // which "copyfiles" does using the "up" option.
+                        const depth = directoryToCopy.split(/[\\/]/).length;
+
+                        // Promisified version of: https://github.com/calvinmetcalf/copyfiles#programic-api.
+                        return copyfiles(sourceAndTargetPaths, { verbose: verbose, up: depth }, () => { });
+                    },
+                    () => process.stderr.write(
+                        `The directory "${directoryToCopy}" cannot be accessed to copy files from.\n`));
+        }))
+        .reduce((previousArray, currentArray) => [...previousArray, ...currentArray], []));
+}
+
+(async function main() {
+    logLine('Executing copy-assets.js...');
+
+    try {
+        await getAssetsConfig({ directory: process.cwd(), verbose: verbose })
+            .then((config) => (config ? copyFilesFromConfig(config) : Promise.resolve()))
+            .then(() => { logLine('Finished executing copy-assets.js.'); });
```

Comment:
It just occurred to me, but why are you mixing async and then chains?

Reply:
Probably because I'm still learning these concepts and don't have any experience in writing `async` code. But I've also found code samples that use both together, so it didn't feel that bad.

Reply:
Maybe they had some good reason, but here it's unnecessary complication. You can replace the wrapping try-catch with a final `.catch()` at the end of the chain and replace the `await` with a `return`. Or you can await each of these 3 lines individually. The end result is the same.

Reply:
You have the same in _clean-assets.js_ too.

https://github.com/Lombiq/NodeJs-Extensions/blob/52549353b70d4eee3daa87cd890228ac37374883/Lombiq.NodeJs.Extensions/scripts/clean-assets.js#L35-L37